pallets/polkadex-ido/src/lib.rs

```diff
@@ -843,52 +383,38 @@ pub mod pallet {
-            // Check if there is any left to withdraw
-            let total_tokens_bought_by_investors = if T::OnePDEX::get().saturated_into::<BalanceOf<T>>() >= funding_round.token_a_priceper_token_b {
-                funding_round.token_a_price_per_1e12_token_b().saturating_reciprocal_mul(funding_round.amount)
-            } else {
-                funding_round.amount / funding_round.token_a_price_per_1e12_token_b_balance()   //TODO saturated div
-            };
-            let remaining_token = funding_round.amount.saturating_sub(total_tokens_bought_by_investors);
+            let funding_round = <InfoFundingRound<T>>::get(round_id).ok_or(Error::<T>::FundingRoundNotApproved)?;
             ensure!(current_block_no >= funding_round.close_round_block, Error::<T>::WithdrawalBlocked);
-            ensure!(remaining_token > Zero::zero(), Error::<T>::WithdrawalBlocked);
             let round_account_id = Self::round_account_id(round_id.clone());
-            //Transfers to remaining token back to creator after round.
-            Self::transfer(funding_round.token_a, &round_account_id, &beneficiary, remaining_token.saturated_into())?;
-            Self::deposit_event(Event::WithdrawToken(round_id, creator));
+            let rem_tokens = funding_round.amount.saturating_sub(funding_round.actual_raise);
+            Self::transfer(funding_round.token_a, &round_account_id, &investor_address, rem_tokens.saturated_into())?;
             Ok(())
         }
-    }
-
-    /// Stores investor Info
-    #[pallet::storage]
-    #[pallet::getter(fn get_investorinfo)]
-    pub(super) type InfoInvestor<T: Config> = StorageMap<
-        _,
-        Blake2_128Concat,
-        T::AccountId,
-        InvestorInfo<T>,
-        OptionQuery,
-    >;
 
+        #[pallet::weight((10_000, DispatchClass::Normal))]
+        pub fn withdraw_investment(origin: OriginFor<T>, round_id: T::Hash, beneficiary: T::AccountId) -> DispatchResult {
+            let investor_address: T::AccountId = ensure_signed(origin)?;
+            ensure!(<InfoFundingRound<T>>::contains_key(&round_id.clone()), Error::<T>::FundingRoundDoesNotExist);
+            let current_block_no = <frame_system::Pallet<T>>::block_number();
+            let funding_round = <InfoFundingRound<T>>::get(round_id).ok_or(Error::<T>::FundingRoundNotApproved)?;
+            ensure!(current_block_no >= funding_round.close_round_block, Error::<T>::WithdrawalBlocked);
+            let total_raise: u128 = funding_round.actual_raise.saturated_into();
+            let amount: u128 = funding_round.amount.saturated_into();
+            ensure!(total_raise <= amount/2, Error::<T>::CannotWithdrawForSuccesfulIDO);
+            let round_account_id = Self::round_account_id(round_id.clone());
+            let investment = <InvestorInvestment<T>>::get(&round_id, &investor_address);
+            Self::transfer(funding_round.token_b, &round_account_id, &investor_address, investment)?;
```

Comment:
Shouldn't `InvestorInvestment` be updated after this? Or it will be done within transfer method?

Reply:
Updated ✅